File: backend-python/app/services/calibration.py
CAPACITY_MAP: dict[str, float] = {
    # JG-B3 options
    "nunca lo he hecho": 1,
    "he leído sobre ello": 3.3,
    "lo he practicado poco": 6.6,
    "lo domino": 10,
    # JG-M3 options
    "entender qué me piden": 2.5,
    "encontrar la información necesaria": 5,
    "organizar mis ideas": 7.5,
    "redactar o presentar el resultado": 10,
    # JG-A3 options
    "no tengo experiencia previa relevante": 1,
    "tengo algo de experiencia pero no estoy seguro/a de aplicarla": 3.3,
    "mi experiencia previa me da buena base": 6.6,
    "mi experiencia previa me prepara completamente": 10,
    # Legacy / generic
    "muy capaz": 10,
    "capaz": 8,
    "algo capaz": 6,
    "poco capaz": 4,
    "no muy capaz": 3,
    "no me siento capaz": 1,
    "no capaz": 0,
}

VALID_CAPACITY_VALUES = set(CAPACITY_MAP.keys())


def normalize_jol(jol: dict) -> float:
    jol_type = jol["tipo"]
    value = jol["valor"]

    if jol_type == "escala":
        confidence = float(value)
        if confidence > 10:
            confidence = confidence / 10.0
    elif jol_type == "tiempo":
        max_time = float(jol.get("tiempo_maximo", 15))
        confidence = 10.0 - (float(value) / max_time) * 10.0
    elif jol_type == "capacidad":
        if isinstance(value, str):
            confidence = CAPACITY_MAP.get(value.lower().strip(), 5.0)
        else:
            confidence = float(value)
    else:
        raise ValueError(f"Unsupported JOL type: {jol_type}")

    return max(0.0, min(10.0, confidence))


def normalize_performance(value: float) -> float:
    if value > 10:
        return value / 10.0
    return value


def compute_predicted_confidence(jols: list[dict]) -> dict:
    confidences = [normalize_jol(j) for j in jols]
    average = sum(confidences) / len(confidences) if confidences else 0.0

    return {
        "confianzas_normalizadas": [round(c, 2) for c in confidences],
        "nivel_confianza_predicho": round(average, 2),
        "etiqueta": _classify_confidence(average),
    }


def compute_calibration(
    jols: list[dict],
    actual_scores: list[float],
) -> dict:
    if len(jols) != len(actual_scores):
        raise ValueError("JOLs and actual scores must have the same length")

    if not jols:
        return {
            "confianzas_normalizadas": [],
            "resultados_normalizados": [],
            "diferencias_por_item": [],
            "error_promedio": 0.0,
            "calibracion": 5.0,
            "bias_index": 0.0,
            "cluster": "cal",
            "nivel": "Sin suficientes observaciones",
        }

    confidences = [normalize_jol(j) for j in jols]
    performances = [normalize_performance(s) for s in actual_scores]

    avg_c = sum(confidences) / len(confidences)
    avg_p = sum(performances) / len(performances)

    diffs = [abs(c - p) for c, p in zip(confidences, performances)]
    avg_error = sum(diffs) / len(diffs)

    bias_index = round(avg_c - avg_p, 2)
    calibration = max(0.0, min(10.0, 10.0 - avg_error))

    # Cluster determination based on Bias Index
    if bias_index > 2.0:
        cluster = "over"
    elif bias_index < -2.0:
        cluster = "sub"
    else:
        cluster = "cal"

    # Edge case: zero performance and zero confidence (0,0)
    is_zero_attempt = (avg_c < 1.0 and avg_p < 1.0)
    nivel = _classify_calibration(calibration, bias_index, is_zero_attempt)

    return {
        "confianzas_normalizadas": [round(c, 2) for c in confidences],
        "resultados_normalizados": [round(p, 2) for p in performances],
        "diferencias_por_item": [round(d, 2) for d in diffs],
        "error_promedio": round(avg_error, 2),
        "bias_index": bias_index,
        "calibracion": round(calibration, 2),
        "cluster": cluster,
        "nivel": nivel,
        "is_zero_attempt": is_zero_attempt,
    }


def _classify_confidence(avg: float) -> str:
    if avg >= 9:
        return "Confianza muy alta"
    elif avg >= 7:
        return "Confianza alta"
    elif avg >= 5:
        return "Confianza media"
    elif avg >= 3:
        return "Confianza baja"
    return "Confianza muy baja"


def _classify_calibration(cal: float, bias: float, is_zero_attempt: bool = False) -> str:
    if is_zero_attempt:
        return "Reconocimiento de falta de dominio (Desempeño Nulo)"

    if bias > 2.0:
        return f"Sobreconfianza Marcada (Sesgo +{bias})"
    elif bias < -2.0:
        return f"Subestimación Cognitiva (Sesgo {bias})"

    if cal >= 9:
        return "Excelente Calibración Metacognitiva"
    elif cal >= 7:
        return "Buena Calibración"
    elif cal >= 5:
        return "Calibración Regular"
    return "Baja Calibración Metacognitiva"

File: frontend/src/lib/calibration.ts
export type JolType = 'escala' | 'tiempo' | 'capacidad';

export interface JolInput {
  tipo: JolType;
  valor: number | string;
  tiempo_maximo?: number;
}

const CAPACITY_MAP: Record<string, number> = {
  'nunca lo he hecho': 1,
  'he leído sobre ello': 3.3,
  'lo he practicado poco': 6.6,
  'lo domino': 10,
  'entender qué me piden': 2.5,
  'encontrar la información necesaria': 5,
  'organizar mis ideas': 7.5,
  'redactar o presentar el resultado': 10,
  'no tengo experiencia previa relevante': 1,
  'tengo algo de experiencia pero no estoy seguro/a de aplicarla': 3.3,
  'mi experiencia previa me da buena base': 6.6,
  'mi experiencia previa me prepara completamente': 10,
  'muy capaz': 10,
  capaz: 8,
  'algo capaz': 6,
  'poco capaz': 4,
  'no muy capaz': 3,
  'no me siento capaz': 1,
  'no capaz': 0,
};

export function normalizeJol(jol: JolInput): number {
  const { tipo, valor: raw, tiempo_maximo } = jol;

  let confidence: number;

  if (tipo === 'escala') {
    confidence = raw as number;
    if (confidence > 10) confidence = confidence / 10;
  } else if (tipo === 'tiempo') {
    const maxTime = tiempo_maximo ?? 15;
    confidence = 10 - ((raw as number) / maxTime) * 10;
  } else if (tipo === 'capacidad') {
    if (typeof raw === 'string') {
      const numMatch = raw.match(/^(\d+)=/);
      if (numMatch) {
        const n = parseInt(numMatch[1]);
        const maxOpts = 4;
        confidence = ((n - 1) / (maxOpts - 1)) * 10;
      } else {
        confidence = CAPACITY_MAP[raw.toLowerCase().trim()] ?? 5;
      }
    } else {
      confidence = raw as number;
    }
  } else {
    throw new Error(`Unsupported JOL type: ${tipo}`);
  }

  return Math.max(0, Math.min(10, confidence));
}

export function normalizePerformance(value: number): number {
  if (value > 10) return value / 10;
  return value;
}

export function computePredictedConfidence(jols: JolInput[]): {
  confianzas_normalizadas: number[];
  nivel_confianza_predicho: number;
  etiqueta: string;
} {
  const confidences = jols.map(normalizeJol);
  const avg = confidences.reduce((a, b) => a + b, 0) / confidences.length;

  return {
    confianzas_normalizadas: confidences.map((c) => Math.round(c * 100) / 100),
    nivel_confianza_predicho: Math.round(avg * 100) / 100,
    etiqueta: classifyConfidence(avg),
  };
}

export function computeCalibration(
  jols: JolInput[],
  actualScores: number[],
): {
  confianzas_normalizadas: number[];
  resultados_normalizados: number[];
  diferencias_por_item: number[];
  error_promedio: number;
  calibracion: number;
  nivel: string;
} {
  if (jols.length !== actualScores.length) {
    throw new Error('JOLs and actual scores must have the same length');
  }

  const confidences = jols.map(normalizeJol);
  const performances = actualScores.map(normalizePerformance);

  const diffs = confidences.map((c, i) => Math.abs(c - performances[i]));
  const avgError = diffs.reduce((a, b) => a + b, 0) / diffs.length;
  const calibration = Math.max(0, Math.min(10, 10 - avgError));

  return {
    confianzas_normalizadas: confidences.map((c) => Math.round(c * 100) / 100),
    resultados_normalizados: performances.map((p) => Math.round(p * 100) / 100),
    diferencias_por_item: diffs.map((d) => Math.round(d * 100) / 100),
    error_promedio: Math.round(avgError * 100) / 100,
    calibracion: Math.round(calibration * 100) / 100,
    nivel: classifyCalibration(calibration),
  };
}

function classifyConfidence(avg: number): string {
  if (avg >= 9) return 'Confianza muy alta';
  if (avg >= 7) return 'Confianza alta';
  if (avg >= 5) return 'Confianza media';
  if (avg >= 3) return 'Confianza baja';
  return 'Confianza muy baja';
}

function classifyCalibration(cal: number): string {
  if (cal >= 9) return 'Excelente calibración';
  if (cal >= 7) return 'Buena calibración';
  if (cal >= 5) return 'Calibración regular';
  if (cal >= 3) return 'Baja calibración (sobreconfianza o subconfianza)';
  return 'Muy baja calibración';
}
